src/NServiceBus.Heartbeat/HeartbeatSender.cs

```diff
@@ -23,41 +22,56 @@ class HeartbeatSender : FeatureStartupTask, IDisposable
             this.hostInfo = hostInfo;
         }
 
-        public void Dispose()
-        {
-            cancellationTokenSource?.Dispose();
-        }
+        public void Dispose() =>
+            stopSendingHeartbeatsTokenSource?.Dispose();
 
-        protected override Task OnStart(IMessageSession session)
+        protected override Task OnStart(IMessageSession session, CancellationToken cancellationToken = default)
         {
-            cancellationTokenSource = new CancellationTokenSource();
-
-            NotifyEndpointStartup(DateTime.UtcNow);
-            StartHeartbeats();
+            stopSendingHeartbeatsTokenSource = new CancellationTokenSource();
 
-            return Task.FromResult(0);
-        }
-
-        protected override Task OnStop(IMessageSession session)
-        {
-            heartbeatTimer?.Stop();
+            // don't block here since StartupTasks are executed synchronously.
+            _ = SendEndpointStartupMessage(DateTime.UtcNow, cancellationToken);
 
-            cancellationTokenSource?.Cancel();
+            Logger.Debug($"Start sending heartbeats every {heartbeatInterval}");
 
-            return Task.FromResult(0);
+            _ = Task.Run(async () =>
+                {
+                    while (!stopSendingHeartbeatsTokenSource.IsCancellationRequested)
+                    {
+                        try
+                        {
+                            await Task.Delay(heartbeatInterval, stopSendingHeartbeatsTokenSource.Token).ConfigureAwait(false);
+
+                            var message = new EndpointHeartbeat
+                            {
+                                ExecutedAt = DateTime.UtcNow,
+                                EndpointName = endpointName,
+                                Host = hostInfo.DisplayName,
+                                HostId = hostInfo.HostId
+                            };
+
+                            await backend.Send(message, ttlTimeSpan, dispatcher, stopSendingHeartbeatsTokenSource.Token).ConfigureAwait(false);
+                        }
+                        catch (OperationCanceledException)
+                        {
+                            // no-op
+                        }
+                        catch (Exception ex)
+                        {
+                            Logger.Warn("Unable to send heartbeat to ServiceControl.", ex);
+                        }
+                    }
+                },
+                cancellationToken);
```

Comment:
We shouldn't pass the token here unless we want to participate in cooperative cancellation in this method, in which case we need to do this — https://github.com/Particular/NServiceBus.RavenDB/blob/371154f9c9293edc728e9dc4b461739f37a4e6e3/src/NServiceBus.RavenDB/Outbox/RavenDbOutboxStorage.cs#L69-L72

Reply:
Discussed with @DavidBoike and will change to `CancellationToken.None`.